frontend/src/features/toy/ToyList.tsx

```diff
@@ -5,67 +5,100 @@ import { RootState } from "../../store/store";
 import { fetchToys, setPage, setResultsPerPage } from "./toySlice";
 import ToyCard from "./ToyCard";
 import Pagination from "../../components/Pagination";
+import Error from "../../components/Error";
+import Loader from "../../components/Loader";
+import ToyFilterList from "./Filter/ToyFilterList";
 
 function ToyList() {
```

Comment:
1. **Code Structure**:
   - Current: The fetchToys call is made inside useEffect.
   - Improvement: The dispatch logic has been optimized and redundancy reduced by extracting the fetchToysData function. This makes the useEffect hook cleaner and more readable.

2. **Fetching Toys**:
   - Current: The dispatch call includes parameters directly inside the useEffect hook.
   - Improvement: A params object is created and passed to the fetchToys function, improving readability and making it easier to manage parameters.

3. **Loading and Error Handling**:
   - Improvement: Loading and error states are handled at the beginning of the component, returning the Loader and Error components respectively. This reduces nesting and improves readability.

4. **Pagination and Toy Rendering**:
   - Improvement: The Pagination component is rendered both above and below the ToyCard list to enhance user experience by providing easy access to pagination controls. The logic for rendering ToyCard components is more concise.

5. **Dispatch Calls**:
   - Improvement: The dispatch calls for setPage and setResultsPerPage are simplified by using inline arrow functions, reducing verbosity.

```typescript
function ToyList() {
  const dispatch = useAppDispatch();
  const { toys, error, loading, currentPage, totalPages, resultsPerPage, totalResults, ageCategory, brand, category, sort, search } = useAppSelector(state => state.toys);

  useEffect(() => {
    const fetchToysData = () => {
      const params = {
        page: currentPage,
        limit: resultsPerPage,
        category: category.join(","),
        ageCategory: ageCategory.join(","),
        brand: brand.join(","),
        sort,
        search
      };
      dispatch(fetchToys(params));
    };

    fetchToysData();
  }, [dispatch, currentPage, resultsPerPage, category, ageCategory, brand, sort, search]);

  if (loading) return <Loader />;
  if (error) return <Error errorMessage={error ?? "Something went wrong.."} />;

  return (
    <div className="min-h-[100%] relative">
      <section className="flex gap-5 px-4 sm:px-10">
        <ToyFilterList />
        <section>
          <Pagination
            currentPage={currentPage}
            totalPages={totalPages}
            onPageChange={page => dispatch(setPage(page))}
            resultsPerPage={resultsPerPage}
            onResultsPerPageChange={results => {
              dispatch(setPage(1));
              dispatch(setResultsPerPage(results));
            }}
            totalResults={totalResults}
          />
          <div className="flex flex-wrap justify-center ">
            {toys.map(toy => (
              <ToyCard
                key={toy._id}
                id={toy._id}
                name={toy.name}
                description={toy.description}
                images={toy.images.length > 0 ? toy.images : ["../bear.webp"]}
                tokens={toy.tokenValue}
              />
            ))}
          </div>
          <Pagination
            currentPage={currentPage}
            totalPages={totalPages}
            onPageChange={page => dispatch(setPage(page))}
            resultsPerPage={resultsPerPage}
            onResultsPerPageChange={results => dispatch(setResultsPerPage(results))}
            totalResults={totalResults}
          />
        </section>
      </section>
    </div>
  );
}